fabric-data-generation-api-v1/src/client/java/net/fabricmc/fabric/api/client/datagen/v1/provider/FabricSoundsProvider.java

```diff
@@ -0,0 +1,119 @@
+ */
+
+package net.fabricmc.fabric.api.client.datagen.v1.provider;
+
+import java.nio.file.Path;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+import com.mojang.serialization.Codec;
+import org.jetbrains.annotations.ApiStatus;
+
+import net.minecraft.data.DataOutput;
+import net.minecraft.data.DataProvider;
+import net.minecraft.data.DataWriter;
+import net.minecraft.registry.RegistryWrapper;
+import net.minecraft.registry.entry.RegistryEntry;
+import net.minecraft.sound.SoundEvent;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.client.datagen.v1.builder.SoundTypeBuilder;
+import net.fabricmc.fabric.api.datagen.v1.FabricDataGenerator;
+import net.fabricmc.fabric.impl.datagen.client.SoundTypeBuilderImpl;
+
+/**
+ * Extend this class and implement {@link FabricSoundsProvider#generate}.
+ *
+ * <p>Register an instance of the class with {@link FabricDataGenerator.Pack#addProvider} in a {@link net.fabricmc.fabric.api.datagen.v1.DataGeneratorEntrypoint}.
+ *
+ * <p>Registered sound types will be appended to their own {@code sounds.json} in a namespace corresponding to
+ * the id of the sound event they are assigned to.
+ */
+public abstract class FabricSoundsProvider implements DataProvider {
+	private static final Codec<Map<String, SoundTypeBuilderImpl.SoundType>> CODEC = Codec.unboundedMap(Codec.STRING, SoundTypeBuilderImpl.SoundType.CODEC);
+	private final CompletableFuture<RegistryWrapper.WrapperLookup> registriesFuture;
+	private final DataOutput output;
+
+	public FabricSoundsProvider(DataOutput output, CompletableFuture<RegistryWrapper.WrapperLookup> registriesFuture) {
+		this.registriesFuture = registriesFuture;
+		this.output = output;
+	}
+
+	@Override
+	public CompletableFuture<?> run(DataWriter writer) {
+		return registriesFuture.thenCompose(lookup -> {
+			final Map<String, Map<String, SoundTypeBuilderImpl.SoundType>> data = new LinkedHashMap<>();
+			configure((id, builder) -> {
+				if (data.computeIfAbsent(id.getNamespace(), n -> new LinkedHashMap<>()).put(id.getPath(), ((SoundTypeBuilderImpl) builder).build()) != null) {
+					throw new IllegalStateException("Duplicate sound for event " + id);
+				}
+			});
+
+			return CompletableFuture.allOf(data.entrySet().stream().map(file -> {
+				Path outputPath = output.resolvePath(DataOutput.OutputType.RESOURCE_PACK).resolve(file.getKey() + "/sounds.json");
+				return DataProvider.writeCodecToPath(writer, lookup, CODEC, file.getValue(), outputPath);
+			}).toArray(CompletableFuture[]::new));
+		});
+	}
+
+	/**
+	 * Implement this method and then use {@link BiConsumer#accept} to register sound events to be data-generated.
+	 *
+	 * <p>Registered sound types will be appended to their own sounds.json in a namespace corresponding to
+	 * the id of the sound event they are assigned to.
+	 */
+	protected abstract void configure(SoundExporter exporter);
```

Comment:
Might be handy to pass the wrapper lookup, im sure someone will find a use for it.